server/datastore/mysql/migrations/tables/20220323152301_CleanupHostRelatedTables.go

```diff
@@ -0,0 +1,51 @@
+package tables
+
+import (
+	"database/sql"
+	"fmt"
+
+	"github.com/pkg/errors"
+)
+
+func init() {
+	MigrationClient.AddMigration(Up_20220323152301, Down_20220323152301)
+}
+
+func Up_20220323152301(tx *sql.Tx) error {
+	var hostRefs = []string{
+		"host_seen_times",
+		"host_software",
+		"host_users",
+		"host_emails",
+		"host_additional",
+		"scheduled_query_stats",
+		"label_membership",
+		"policy_membership",
+		"host_mdm",
+		"host_munki_info",
+		"host_device_auth",
+	}
+
+	const delStmt = `
+    DELETE target
+    FROM %s target
+    WHERE NOT EXISTS (
+      SELECT 1
+      FROM
+        hosts h
+      WHERE
+        h.id = target.host_id
+    )`
```

Comment:
To answer my own question, I used `EXPLAIN` on each of these queries and I think your original query is best (particularly the 1 row examined on the dependent subquery in your original)

original:
```
+----+--------------------+--------+------------+--------+---------------+---------+---------+----------------------+--------+----------+-------------+
| id | select_type        | table  | partitions | type   | possible_keys | key     | key_len | ref                  | rows   | filtered | Extra       |
+----+--------------------+--------+------------+--------+---------------+---------+---------+----------------------+--------+----------+-------------+
|  1 | DELETE             | target | NULL       | ALL    | NULL          | NULL    | NULL    | NULL                 | 151830 |   100.00 | Using where |
|  2 | DEPENDENT SUBQUERY | h      | NULL       | eq_ref | PRIMARY       | PRIMARY | 4       | fleet.target.host_id |      1 |   100.00 | Using index |
+----+--------------------+--------+------------+--------+---------------+---------+---------+----------------------+--------+----------+-------------+
```

proposed change:
```
+----+-------------+--------+------------+-------+---------------+------------------+---------+------+--------+----------+-------------+
| id | select_type | table  | partitions | type  | possible_keys | key              | key_len | ref  | rows   | filtered | Extra       |
+----+-------------+--------+------------+-------+---------------+------------------+---------+------+--------+----------+-------------+
|  1 | DELETE      | target | NULL       | ALL   | NULL          | NULL             | NULL    | NULL | 151830 |   100.00 | Using where |
|  2 | SUBQUERY    | hosts  | NULL       | index | PRIMARY       | fk_hosts_team_id | 5       | NULL |   5652 |   100.00 | Using index |
+----+-------------+--------+------------+-------+---------------+------------------+---------+------+--------+----------+-------------+
```

Reply:
My understanding of this is that `NOT EXISTS (SELECT ...)` will stop the sub-select as soon as it finds 1 row, while `NOT IN (SELECT ...)` must build the full result set. That being said, I find DB optimizations are hard to predict, good idea to validate with the plan. In this scenario, one is a "dependent subquery" (executed for each potential row to delete) while the other is just a "subquery", so it could be that the full result set of the subquery is built only once and reused for all rows, so again not easy to say (for me at least) which would be faster, but I think it's likely that both would be fast given the use of a straightforward index.